{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb527364-100c-45f2-a640-90f74d39802d",
   "metadata": {},
   "source": [
    "<h1> Quantifying Correlations and Making Figures </h1> \n",
    "<p> This python notebook contains the code we used to quantify the huge number of correlation values, note other analyses were done at the same time and these are also included. This notebook starts with the scripts used to generate the figures within the main text. There is commenting throughout to guide if interested. Then the actual analysis used on the computer cluster is toward the end. This is mostly intuative. -- Bohrer 2024 </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d1ed60ef-a1a8-49a5-b241-384a69ce33e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from npeet import entropy_estimators as ee\n",
    "import os\n",
    "from multiprocessing import Pool\n",
    "import numpy as np\n",
    "import pickle \n",
    "pickle.HIGHEST_PROTOCOL = 4\n",
    "import pandas as pd\n",
    "from scipy.stats import norm, gaussian_kde\n",
    "from scipy.spatial.distance import pdist\n",
    "from scipy.spatial.distance import cdist\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import clear_output\n",
    "import time\n",
    "import copy\n",
    "import seaborn as sns\n",
    "import dill #This will be used to store the workplace so you dont have to run certain cells\n",
    "sns.set()\n",
    "from scipy.stats import entropy, norm\n",
    "from pathlib import Path\n",
    "from math import nan\n",
    "import scipy\n",
    "from scipy.misc import derivative\n",
    "from scipy.integrate import quad\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a858b2d-9692-481d-b40f-ddfa78fa628a",
   "metadata": {},
   "source": [
    "<h2> Repeat one, no constraints </h2> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c0846584-aa0b-45a7-9f2b-d69b20af728d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Im in\n"
     ]
    },
    {
     "ename": "EmptyDataError",
     "evalue": "No columns to parse from file",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmptyDataError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/kt/fz3jn40j3fs94gf_7pfhx25wb9vwf6/T/ipykernel_75820/1735971852.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    153\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    154\u001b[0m     \u001b[0;31m#Lets load in the needed data:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 155\u001b[0;31m     \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mData_Folder\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mSpecific_file_1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mskiprows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    156\u001b[0m     \u001b[0mdf2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_table\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mData_Folder\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mSpecific_file_2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mskiprows\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/util/_decorators.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    309\u001b[0m                     \u001b[0mstacklevel\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mstacklevel\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m                 )\n\u001b[0;32m--> 311\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    313\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mwrapper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36mread_table\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, error_bad_lines, warn_bad_lines, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options)\u001b[0m\n\u001b[1;32m    775\u001b[0m     \u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds_defaults\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    776\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 777\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    778\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    779\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    573\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    574\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 575\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    577\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    930\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    931\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mIOHandles\u001b[0m \u001b[0;34m|\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 932\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    933\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    934\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/readers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1232\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1233\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1234\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mmapping\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1235\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1236\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhandles\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/io/parsers/c_parser_wrapper.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"dtype\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mensure_dtype_objs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"dtype\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 75\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     76\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     77\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.9/site-packages/pandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mEmptyDataError\u001b[0m: No columns to parse from file"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Because there is so much data, we have to go through and make each line of the plot individually so that we dont run out of memory, ugh\n",
    "fig=plt.figure()\n",
    "\n",
    "#This is the metric we are going to use, we are interested in the spearman\n",
    "Correlation_Metric_1='Spear Control'\n",
    "Correlation_Metric_2='Spear Pert'\n",
    "\n",
    "#What we are going to show in the end\n",
    "Calculation='Control Spear-Perterbation Spear'\n",
    "\n",
    "#Because this is going to be slow we are going to do one condistion at a time.\n",
    "Data_Folder='Data_Analyzed/03072023/'\n",
    "\n",
    "for pick in range(0, 5):\n",
    "\n",
    "    #Here I am organizing a giant dataframe and taking into consideration each of the datasets\n",
    "    df_temper={'Condition':[], Calculation : nan}\n",
    "    result2=pd.DataFrame(df_temper)\n",
    "    \n",
    "    Label='RAD21'\n",
    "    condition='RAD21_ALL_1_'\n",
    "    condition_filename='RAD21d_T120/'\n",
    "    Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "    Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    Label='JQ1'\n",
    "    condition='JQ1_ALL_1_'\n",
    "    condition_filename='JQ1_T120/'\n",
    "    Specific_file_1='JQ1_T120/HCT_DMSOSr1.txt'\n",
    "    Specific_file_2='JQ1_T120/HCT_JQ1Sr1.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    Label='MED'\n",
    "    condition='MED_All_1_'\n",
    "    condition_filename='MED26d_T120/'\n",
    "    Specific_file_1='MED26d_T120/HCTMED26D_DMSOr1M.txt'\n",
    "    Specific_file_2='MED26d_T120/HCTMED26D_IAAr1M.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "    Label='Control JQ1 R1 vs. Control JQ1 R2'\n",
    "    condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "    condition_filename='JQ1_T120_Controls/'\n",
    "    Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "    Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "\n",
    "    Label='Control MED R1 vs. Control MED R2'\n",
    "    condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "    condition_filename='MED26d_T120_Controls/'\n",
    "    Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "    Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "    \n",
    "    #Now load the data for real\n",
    "    \n",
    "    if pick==0:\n",
    "        Label='RAD21'\n",
    "        condition='RAD21_ALL_1_'\n",
    "        condition_filename='RAD21d_T120/'\n",
    "        Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "        Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "   \n",
    "    if pick==1:\n",
    "        Label='JQ1'\n",
    "        condition='JQ1_ALL_1_'\n",
    "        condition_filename='JQ1_T120/'\n",
    "        Specific_file_1='JQ1_T120/HCT_DMSOSr1.txt'\n",
    "        Specific_file_2='JQ1_T120/HCT_JQ1Sr1.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    " \n",
    "        #Now for Med\n",
    "    if pick==2:\n",
    "        Label='MED'\n",
    "        condition='MED_All_1_'\n",
    "        condition_filename='MED26d_T120/'\n",
    "        Specific_file_1='MED26d_T120/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED26d_T120/HCTMED26D_IAAr1M.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "        \n",
    "    if pick==3:\n",
    "        Label='Control JQ1 R1 vs. Control JQ1 R2'\n",
    "        condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "        condition_filename='JQ1_T120_Controls/'\n",
    "        Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "        Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    if pick==4:\n",
    "        Label='Control MED R1 vs. Control MED R2'\n",
    "        condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "        condition_filename='MED26d_T120_Controls/'\n",
    "        Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "    if pick==5:\n",
    "        Label='Control MED R1 vs. Control JQ1 R1'\n",
    "        condition='MED_R1_Control_to_JQ1_R1_Control_'\n",
    "        condition_filename='MED_and_JQ1_Controls/'\n",
    "        Specific_file_1='MED_and_JQ1_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED_and_JQ1_Controls/HCT_DMSOSr1.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "\n",
    "    print('Im in')\n",
    "\n",
    "\n",
    "    #Lets load in the needed data: \n",
    "    df = pd.read_table(Data_Folder+Specific_file_1, skiprows=1, header=None)\n",
    "    df2 = pd.read_table(Data_Folder+Specific_file_2, skiprows=1, header=None)\n",
    "\n",
    "    #Here we are going to normalize them the correct way...Reads per million\n",
    "    df=df.set_index(0)\n",
    "    df2=df2.set_index(0)\n",
    "    print(Label, np.mean(np.nansum(df.values, axis=0)))\n",
    "    print(Label, np.mean(np.nansum(df2.values, axis=0)))\n",
    "    \n",
    "    #Here we are going to normalize so that \n",
    "    Lowest_Expression_State=29347\n",
    "    df=df/np.nansum(df.values, axis=0)*Lowest_Expression_State\n",
    "    df2=df2/np.nansum(df2.values, axis=0)*Lowest_Expression_State\n",
    "\n",
    "    df=df.round(0)\n",
    "    df2=df2.round(0)\n",
    "\n",
    "    #We are going to only look at genes that are expressed in at least 10% of the cells\n",
    "    genes_to_keep_control=df.index[np.nansum(df.values>0, axis=1)>len(df.iloc[0])/10]\n",
    "    genes_to_keep_perterb=df2.index[np.nansum(df2.values>0, axis=1)>len(df2.iloc[0])/10]\n",
    "    print(len(genes_to_keep_perterb))\n",
    "\n",
    "    #Here I am going to get rid of the genes that are not in both samples\n",
    "    values_to_kill=np.array([])\n",
    "    for kcat in df.index:\n",
    "        if kcat not in genes_to_keep_control.values:\n",
    "            values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "    df =df[~df.index.isin(values_to_kill)] \n",
    "    df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "\n",
    "\n",
    "    values_to_kill=np.array([])\n",
    "    for kcat in df2.index:\n",
    "        if kcat not in genes_to_keep_perterb.values:\n",
    "            values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "    df =df[~df.index.isin(values_to_kill)]\n",
    "    df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "    \n",
    "    counter=0\n",
    "    for i in df.index.values:\n",
    "        \n",
    "        try:\n",
    "            file=Data_Folder+condition_filename+'Analyzed_Data/'+condition+i+'.csv'\n",
    "            data=pd.read_csv(file)\n",
    "\n",
    "            temper=data[Correlation_Metric_1].values-data[Correlation_Metric_2].values\n",
    "\n",
    "            strs = [Label for x in range(len(temper))]\n",
    "            df_temper={'Condition':strs, Calculation : temper}\n",
    "            df_temper=pd.DataFrame(df_temper)\n",
    "            result2 = pd.concat([result2, df_temper], axis=0)\n",
    "            \n",
    "        except:\n",
    "            skip_this_one=1\n",
    "            \n",
    "    sns.violinplot(data=result2, x=Calculation, y='Condition',inner = \"quartile\")\n",
    "    print('The MAD is the following:')\n",
    "    print(np.nanmedian(np.abs(result2['Control Spear-Perterbation Spear']-np.nanmedian(result2['Control Spear-Perterbation Spear']))))\n",
    "    \n",
    "plt.xlim(-.05,.05)\n",
    "plt.xticks(np.arange(-.05,.05,.02))\n",
    "plt.savefig('Sign_R1_Spear_Diff.pdf',bbox_inches='tight')\n",
    "#plt.savefig('lR1_Spear_Diff_less_than_1_per_change.pdf',bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c3e32a4-c5ee-4c54-9ed7-2fb6b3581d12",
   "metadata": {},
   "source": [
    "<h2> Repeat 2, no constraints </h2> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31188404-0e7b-4b38-b337-aee0861d594d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Because there is so much data, we have to go through and make each line of the plot individually so that we dont run out of memory, ugh\n",
    "fig=plt.figure()\n",
    "\n",
    "#This is the metric we are going to use, we are interested in the spearman\n",
    "Correlation_Metric_1='Spear Control'\n",
    "Correlation_Metric_2='Spear Pert'\n",
    "\n",
    "#What we are going to show in the end\n",
    "Calculation='Control Spear-Perterbation Spear'\n",
    "\n",
    "#Because this is going to be slow we are going to do one condistion at a time.\n",
    "Data_Folder='Data_Analyzed/03072023/'\n",
    "\n",
    "\n",
    "for pick in range(0, 5):\n",
    "    \n",
    "\n",
    "    df_temper={'Condition':[], Calculation : nan}\n",
    "    result2=pd.DataFrame(df_temper)\n",
    "\n",
    "\n",
    "    Label='RAD21'\n",
    "    condition='RAD21_ALL_1_'\n",
    "    condition_filename='RAD21d_T120/'\n",
    "    Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "    Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "\n",
    "\n",
    "    Label='JQ1'\n",
    "    condition='JQ1_ALL_2_'\n",
    "    condition_filename='JQ1_T120/'\n",
    "    Specific_file_1='JQ1_T120/HCT_DMSOSr2.txt'\n",
    "    Specific_file_2='JQ1_T120/HCT_JQ1Sr2.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "\n",
    "   \n",
    "\n",
    "    Label='MED'\n",
    "    condition='MED_All_1_'\n",
    "    condition_filename='MED26d_T120/'\n",
    "    Specific_file_1='MED26d_T120/HCTMED26D_DMSOr2M.txt'\n",
    "    Specific_file_2='MED26d_T120/HCTMED26D_IAAr2M.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "    \n",
    "    \n",
    "     \n",
    "    Label='Control JQ1 R1 vs. Control JQ1 R2'\n",
    "    condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "    condition_filename='JQ1_T120_Controls/'\n",
    "    Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "    Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "\n",
    "    Label='Control MED R1 vs. Control MED R2'\n",
    "    condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "    condition_filename='MED26d_T120_Controls/'\n",
    "    Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "    Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "  \n",
    "    \n",
    "    #Now for real\n",
    "    \n",
    "    if pick==0:\n",
    "        Label='RAD21'\n",
    "        condition='RAD21_ALL_1_'\n",
    "        condition_filename='RAD21d_T120/'\n",
    "        Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "        Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "   \n",
    "    if pick==1:\n",
    "        Label='JQ1'\n",
    "        condition='JQ1_ALL_2_'\n",
    "        condition_filename='JQ1_T120/'\n",
    "        Specific_file_1='JQ1_T120/HCT_DMSOSr2.txt'\n",
    "        Specific_file_2='JQ1_T120/HCT_JQ1Sr2.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    " \n",
    "        #Now for Med\n",
    "    if pick==2:\n",
    "        Label='MED'\n",
    "        condition='MED_All_2_'\n",
    "        condition_filename='MED26d_T120/'\n",
    "        Specific_file_1='MED26d_T120/HCTMED26D_DMSOr2M.txt'\n",
    "        Specific_file_2='MED26d_T120/HCTMED26D_IAAr2M.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "        \n",
    "    if pick==3:\n",
    "        Label='Control JQ1 R1 vs. Control JQ1 R2'\n",
    "        condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "        condition_filename='JQ1_T120_Controls/'\n",
    "        Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "        Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    if pick==4:\n",
    "        Label='Control MED R1 vs. Control MED R2'\n",
    "        condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "        condition_filename='MED26d_T120_Controls/'\n",
    "        Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "    if pick==5:\n",
    "        Label='Control MED R1 vs. Control JQ1 R1'\n",
    "        condition='MED_R1_Control_to_JQ1_R1_Control_'\n",
    "        condition_filename='MED_and_JQ1_Controls/'\n",
    "        Specific_file_1='MED_and_JQ1_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED_and_JQ1_Controls/HCT_DMSOSr1.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    \n",
    "    print('Im in')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #Lets load in the needed data: \n",
    "    df = pd.read_table(Data_Folder+Specific_file_1, skiprows=1, header=None)\n",
    "    df2 = pd.read_table(Data_Folder+Specific_file_2, skiprows=1, header=None)\n",
    "\n",
    "    #Here we are going to normalize them the correct way...Reads per million\n",
    "    df=df.set_index(0)\n",
    "    df2=df2.set_index(0)\n",
    "    print(Label, np.mean(np.nansum(df.values, axis=0)))\n",
    "    print(Label, np.mean(np.nansum(df2.values, axis=0)))\n",
    "    \n",
    "    Lowest_Expression_State=29347\n",
    "    df=df/np.nansum(df.values, axis=0)*Lowest_Expression_State\n",
    "    df2=df2/np.nansum(df2.values, axis=0)*Lowest_Expression_State\n",
    "\n",
    "\n",
    "    df=df.round(0)\n",
    "    df2=df2.round(0)\n",
    "\n",
    "    #We are going to only look at genes that are expressed in at least 10% of the cells\n",
    "    genes_to_keep_control=df.index[np.nansum(df.values>0, axis=1)>len(df.iloc[0])/10]\n",
    "    genes_to_keep_perterb=df2.index[np.nansum(df2.values>0, axis=1)>len(df2.iloc[0])/10]\n",
    "    print(len(genes_to_keep_perterb))\n",
    "\n",
    "    #Here I am going to get rid of the genes that are not in both samples\n",
    "    values_to_kill=np.array([])\n",
    "    for kcat in df.index:\n",
    "        if kcat not in genes_to_keep_control.values:\n",
    "            values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "    df =df[~df.index.isin(values_to_kill)] \n",
    "    df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "\n",
    "\n",
    "    values_to_kill=np.array([])\n",
    "    for kcat in df2.index:\n",
    "        if kcat not in genes_to_keep_perterb.values:\n",
    "            values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "    df =df[~df.index.isin(values_to_kill)]\n",
    "    df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "    \n",
    "    counter=0\n",
    "    for i in df.index.values:\n",
    "\n",
    "        try:\n",
    "            file=Data_Folder+condition_filename+'Analyzed_Data/'+condition+i+'.csv'\n",
    "            data=pd.read_csv(file)\n",
    "\n",
    "        \n",
    "            temper=data[Correlation_Metric_1].values-data[Correlation_Metric_2].values\n",
    "\n",
    "            strs = [Label for x in range(len(temper))]\n",
    "            df_temper={'Condition':strs, Calculation : temper}\n",
    "            df_temper=pd.DataFrame(df_temper)\n",
    "            result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "        except:\n",
    "            skip_this_one=1\n",
    "            \n",
    "\n",
    "    print(np.nanmedian(np.abs(result2['Control Spear-Perterbation Spear']-np.nanmedian(result2['Control Spear-Perterbation Spear']))))\n",
    "    sns.violinplot(data=result2, x=Calculation, y='Condition',inner = \"quartile\")\n",
    "    \n",
    "    \n",
    "plt.xlim(-.05,.05)\n",
    "plt.xticks(np.arange(-.05,.05,.02))\n",
    "plt.savefig('Sign_R2_Spear_Diff.pdf',bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12d43e8c-6250-4029-90ed-51461d54d7d8",
   "metadata": {},
   "source": [
    "<h2> Controling for differences in expression level </h2> \n",
    "<p> Here we only investigate the differences in correlation if the mean expression levels dont really change. First we do repeat 1 and then Repeat 2 </p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "121f41b3-02a0-4568-984a-64a624cb41ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Repeat 1\n",
    "#Because there is so much data, we have to go through and make each line of the plot individually so that we dont run out of memory, ugh\n",
    "fig=plt.figure()\n",
    "\n",
    "#This is the metric we are going to use, we are interested in the spearman\n",
    "Correlation_Metric_1='Spear Control'\n",
    "Correlation_Metric_2='Spear Pert'\n",
    "\n",
    "#What we are going to show in the end\n",
    "Calculation='Control Spear-Perterbation Spear'\n",
    "\n",
    "#Because this is going to be slow we are going to do one condistion at a time.\n",
    "Data_Folder='Data_Analyzed/03072023/'\n",
    "\n",
    "for pick in range(0, 5):\n",
    "    \n",
    "\n",
    "    df_temper={'Condition':[], Calculation : nan}\n",
    "    result2=pd.DataFrame(df_temper)\n",
    "\n",
    "    Label='RAD21'\n",
    "    condition='RAD21_ALL_1_'\n",
    "    condition_filename='RAD21d_T120/'\n",
    "    Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "    Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    Label='JQ1'\n",
    "    condition='JQ1_ALL_1_'\n",
    "    condition_filename='JQ1_T120/'\n",
    "    Specific_file_1='JQ1_T120/HCT_DMSOSr1.txt'\n",
    "    Specific_file_2='JQ1_T120/HCT_JQ1Sr1.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    Label='MED'\n",
    "    condition='MED_All_1_'\n",
    "    condition_filename='MED26d_T120/'\n",
    "    Specific_file_1='MED26d_T120/HCTMED26D_DMSOr1M.txt'\n",
    "    Specific_file_2='MED26d_T120/HCTMED26D_IAAr1M.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "    Label='Control JQ1 R1 vs. Control JQ1 R2'\n",
    "    condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "    condition_filename='JQ1_T120_Controls/'\n",
    "    Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "    Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "\n",
    "    Label='Control MED R1 vs. Control MED R2'\n",
    "    condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "    condition_filename='MED26d_T120_Controls/'\n",
    "    Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "    Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    " \n",
    "    #Now for real\n",
    "    if pick==0:\n",
    "        Label='RAD21'\n",
    "        condition='RAD21_ALL_1_'\n",
    "        condition_filename='RAD21d_T120/'\n",
    "        Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "        Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    if pick==1:\n",
    "        Label='JQ1'\n",
    "        condition='JQ1_ALL_1_'\n",
    "        condition_filename='JQ1_T120/'\n",
    "        Specific_file_1='JQ1_T120/HCT_DMSOSr1.txt'\n",
    "        Specific_file_2='JQ1_T120/HCT_JQ1Sr1.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    if pick==2:\n",
    "        Label='MED'\n",
    "        condition='MED_All_1_'\n",
    "        condition_filename='MED26d_T120/'\n",
    "        Specific_file_1='MED26d_T120/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED26d_T120/HCTMED26D_IAAr1M.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "        \n",
    "    if pick==3:\n",
    "        Label='Control JQ1 R1 vs. Control JQ1 R2'\n",
    "        condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "        condition_filename='JQ1_T120_Controls/'\n",
    "        Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "        Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    if pick==4:\n",
    "        Label='Control MED R1 vs. Control MED R2'\n",
    "        condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "        condition_filename='MED26d_T120_Controls/'\n",
    "        Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "    if pick==5:\n",
    "        Label='Control MED R1 vs. Control JQ1 R1'\n",
    "        condition='MED_R1_Control_to_JQ1_R1_Control_'\n",
    "        condition_filename='MED_and_JQ1_Controls/'\n",
    "        Specific_file_1='MED_and_JQ1_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED_and_JQ1_Controls/HCT_DMSOSr1.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "   \n",
    "\n",
    "    print('Im in')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    #Lets load in the needed data: \n",
    "    df = pd.read_table(Data_Folder+Specific_file_1, skiprows=1, header=None)\n",
    "    df2 = pd.read_table(Data_Folder+Specific_file_2, skiprows=1, header=None)\n",
    "\n",
    "    #Here we are going to normalize them the correct way...Reads per million\n",
    "    df=df.set_index(0)\n",
    "    df2=df2.set_index(0)\n",
    "    print(Label, np.mean(np.nansum(df.values, axis=0)))\n",
    "    print(Label, np.mean(np.nansum(df2.values, axis=0)))\n",
    "    \n",
    "    Lowest_Expression_State=29347\n",
    "    df=df/np.nansum(df.values, axis=0)*Lowest_Expression_State\n",
    "    df2=df2/np.nansum(df2.values, axis=0)*Lowest_Expression_State\n",
    "\n",
    "\n",
    "    df=df.round(0)\n",
    "    df2=df2.round(0)\n",
    "\n",
    "    #We are going to only look at genes that are expressed in at least 10% of the cells\n",
    "    genes_to_keep_control=df.index[np.nansum(df.values>0, axis=1)>len(df.iloc[0])/10]\n",
    "    genes_to_keep_perterb=df2.index[np.nansum(df2.values>0, axis=1)>len(df2.iloc[0])/10]\n",
    "    #print(len(genes_to_keep_perterb))\n",
    "\n",
    "    #Here I am going to get rid of the genes that are not in both samples\n",
    "    values_to_kill=np.array([])\n",
    "    for kcat in df.index:\n",
    "        if kcat not in genes_to_keep_control.values:\n",
    "            values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "    df =df[~df.index.isin(values_to_kill)] \n",
    "    df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "\n",
    "\n",
    "    values_to_kill=np.array([])\n",
    "    for kcat in df2.index:\n",
    "        if kcat not in genes_to_keep_perterb.values:\n",
    "            values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "    df =df[~df.index.isin(values_to_kill)]\n",
    "    df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "    \n",
    "    counter=0\n",
    "    diff=0\n",
    "    \n",
    "    print(len(df.index.values))\n",
    "    ss2=0\n",
    "    diff=0\n",
    "    for i in df.index.values:\n",
    "\n",
    "\n",
    "        try:\n",
    "            file=Data_Folder+condition_filename+'Analyzed_Data/'+condition+i+'.csv'\n",
    "            data=pd.read_csv(file)\n",
    "\n",
    "            temper=data[Correlation_Metric_1].values-data[Correlation_Metric_2].values\n",
    "           \n",
    "            ss2=ss2+len(temper)\n",
    "\n",
    "#           #Here I am going to account for the expression levels\n",
    "            ff1=np.abs(data['Exp Gene1 Pert'].values-data['Exp Gene1 Control'].values)/data['Exp Gene1 Control'].values\n",
    "            ff2=np.abs(data['Exp Gene2 Pert'].values-data['Exp Gene2 Control'].values)/data['Exp Gene2 Control'].values\n",
    "            ff1=ff1<.01\n",
    "            ff2=ff2<.01\n",
    "            ffall=ff1*ff2\n",
    "            temper=temper[ffall]\n",
    "            diff=diff+(len(temper))\n",
    "\n",
    "            strs = [Label for x in range(len(temper))]\n",
    "            df_temper={'Condition':strs, Calculation : temper}\n",
    "            df_temper=pd.DataFrame(df_temper)\n",
    "            result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "        except:\n",
    "            \n",
    "            skip_this_one=1\n",
    "\n",
    "    sns.violinplot(data=result2, x=Calculation, y='Condition',inner = \"quartile\")\n",
    "    print(np.nanmedian(np.abs(result2['Control Spear-Perterbation Spear']-np.nanmedian(result2['Control Spear-Perterbation Spear']))))\n",
    "    print(ss2)\n",
    "    print(diff)\n",
    "    \n",
    "\n",
    "plt.xlim(-.05,.05)\n",
    "plt.xticks(np.arange(-.05,.05,.02))\n",
    "plt.savefig('Same_Expression_Sign_R1_Spear_Diff.pdf',bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "621bbc43-754a-47ff-8483-ade070c190f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now for R2 with little to no change in expression level\n",
    "\n",
    "#Because there is so much data, we have to go through and make each line of the plot individually so that we dont run out of memory, ugh\n",
    "fig=plt.figure()\n",
    "\n",
    "#This is the metric we are going to use, we are interested in the spearman\n",
    "Correlation_Metric_1='Spear Control'\n",
    "Correlation_Metric_2='Spear Pert'\n",
    "\n",
    "#What we are going to show in the end\n",
    "Calculation='Control Spear-Perterbation Spear'\n",
    "\n",
    "#Because this is going to be slow we are going to do one condistion at a time.\n",
    "Data_Folder='Data_Analyzed/03072023/'\n",
    "\n",
    "for pick in range(0, 5):\n",
    "    \n",
    "\n",
    "    df_temper={'Condition':[], Calculation : nan}\n",
    "    result2=pd.DataFrame(df_temper)\n",
    "    \n",
    "    Label='RAD21'\n",
    "    condition='RAD21_ALL_1_'\n",
    "    condition_filename='RAD21d_T120/'\n",
    "    Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "    Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    Label='JQ1'\n",
    "    condition='JQ1_ALL_2_'\n",
    "    condition_filename='JQ1_T120/'\n",
    "    Specific_file_1='JQ1_T120/HCT_DMSOSr2.txt'\n",
    "    Specific_file_2='JQ1_T120/HCT_JQ1Sr2.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    Label='MED'\n",
    "    condition='MED_All_1_'\n",
    "    condition_filename='MED26d_T120/'\n",
    "    Specific_file_1='MED26d_T120/HCTMED26D_DMSOr2M.txt'\n",
    "    Specific_file_2='MED26d_T120/HCTMED26D_IAAr2M.txt'\n",
    "    df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "    Label='Control JQ1 R1 vs. Control JQ1 R2'\n",
    "    condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "    condition_filename='JQ1_T120_Controls/'\n",
    "    Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "    Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "    Label='Control MED R1 vs. Control MED R2'\n",
    "    condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "    condition_filename='MED26d_T120_Controls/'\n",
    "    Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "    Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "    df_temper={'Condition':[Label], Calculation : nan}\n",
    "    df_temper=pd.DataFrame(df_temper)\n",
    "    result2 = pd.concat([result2, df_temper], axis=0)\n",
    "    \n",
    "    \n",
    "    #Now for real data\n",
    "    if pick==0:\n",
    "        Label='RAD21'\n",
    "        condition='RAD21_ALL_1_'\n",
    "        condition_filename='RAD21d_T120/'\n",
    "        Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "        Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "   \n",
    "    if pick==1:\n",
    "        Label='JQ1'\n",
    "        condition='JQ1_ALL_2_'\n",
    "        condition_filename='JQ1_T120/'\n",
    "        Specific_file_1='JQ1_T120/HCT_DMSOSr2.txt'\n",
    "        Specific_file_2='JQ1_T120/HCT_JQ1Sr2.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    if pick==2:\n",
    "        Label='MED'\n",
    "        condition='MED_All_2_'\n",
    "        condition_filename='MED26d_T120/'\n",
    "        Specific_file_1='MED26d_T120/HCTMED26D_DMSOr2M.txt'\n",
    "        Specific_file_2='MED26d_T120/HCTMED26D_IAAr2M.txt'\n",
    "        df_temper={'Condition':[Label], 'Control MI-Perterbation MI':nan}\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "    if pick==3:\n",
    "        Label='Control JQ1 R1 vs. Control JQ1 R2'\n",
    "        condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "        condition_filename='JQ1_T120_Controls/'\n",
    "        Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "        Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "    if pick==4:\n",
    "        Label='Control MED R1 vs. Control MED R2'\n",
    "        condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "        condition_filename='MED26d_T120_Controls/'\n",
    "        Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "        \n",
    "    if pick==5:\n",
    "        Label='Control MED R1 vs. Control JQ1 R1'\n",
    "        condition='MED_R1_Control_to_JQ1_R1_Control_'\n",
    "        condition_filename='MED_and_JQ1_Controls/'\n",
    "        Specific_file_1='MED_and_JQ1_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "        Specific_file_2='MED_and_JQ1_Controls/HCT_DMSOSr1.txt'\n",
    "        df_temper={'Condition':[Label], Calculation : nan}\n",
    "        df_temper=pd.DataFrame(df_temper)\n",
    "        result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "    \n",
    "\n",
    "    print('Im in')\n",
    "\n",
    "    #Lets load in the needed data: \n",
    "    df = pd.read_table(Data_Folder+Specific_file_1, skiprows=1, header=None)\n",
    "    df2 = pd.read_table(Data_Folder+Specific_file_2, skiprows=1, header=None)\n",
    "\n",
    "    #Here we are going to normalize them the correct way...Reads per million\n",
    "    df=df.set_index(0)\n",
    "    df2=df2.set_index(0)\n",
    "    print(Label, np.mean(np.nansum(df.values, axis=0)))\n",
    "    print(Label, np.mean(np.nansum(df2.values, axis=0)))\n",
    "    \n",
    "    Lowest_Expression_State=29347\n",
    "    df=df/np.nansum(df.values, axis=0)*Lowest_Expression_State\n",
    "    df2=df2/np.nansum(df2.values, axis=0)*Lowest_Expression_State\n",
    "\n",
    "\n",
    "    df=df.round(0)\n",
    "    df2=df2.round(0)\n",
    "\n",
    "    #We are going to only look at genes that are expressed in at least 10% of the cells\n",
    "    genes_to_keep_control=df.index[np.nansum(df.values>0, axis=1)>len(df.iloc[0])/10]\n",
    "    genes_to_keep_perterb=df2.index[np.nansum(df2.values>0, axis=1)>len(df2.iloc[0])/10]\n",
    "    #print(len(genes_to_keep_perterb))\n",
    "\n",
    "    #Here I am going to get rid of the genes that are not in both samples\n",
    "    values_to_kill=np.array([])\n",
    "    for kcat in df.index:\n",
    "        if kcat not in genes_to_keep_control.values:\n",
    "            values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "    df =df[~df.index.isin(values_to_kill)] \n",
    "    df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "\n",
    "\n",
    "    values_to_kill=np.array([])\n",
    "    for kcat in df2.index:\n",
    "        if kcat not in genes_to_keep_perterb.values:\n",
    "            values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "    df =df[~df.index.isin(values_to_kill)]\n",
    "    df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "    \n",
    "    counter=0\n",
    "    \n",
    "    print(len(df.index.values))\n",
    "    ss2=0\n",
    "    diff=0\n",
    "    for i in df.index.values:\n",
    "\n",
    "        try:\n",
    "            file=Data_Folder+condition_filename+'Analyzed_Data/'+condition+i+'.csv'\n",
    "            data=pd.read_csv(file)\n",
    "            \n",
    "            temper=data[Correlation_Metric_1].values-data[Correlation_Metric_2].values\n",
    "            ss2=ss2+len(temper)\n",
    "            \n",
    "            #Here I am going to account for the expression levels\n",
    "            ff1=np.abs(data['Exp Gene1 Pert'].values-data['Exp Gene1 Control'].values)/data['Exp Gene1 Control'].values\n",
    "            ff2=np.abs(data['Exp Gene2 Pert'].values-data['Exp Gene2 Control'].values)/data['Exp Gene2 Control'].values\n",
    "            ff1=ff1<.01\n",
    "            ff2=ff2<.01\n",
    "            ffall=ff1*ff2\n",
    "            temper=temper[ffall]\n",
    "            diff=diff+len(temper)\n",
    "\n",
    "            strs = [Label for x in range(len(temper))]\n",
    "            df_temper={'Condition':strs, Calculation : temper}\n",
    "            df_temper=pd.DataFrame(df_temper)\n",
    "            result2 = pd.concat([result2, df_temper], axis=0)\n",
    "\n",
    "        except:\n",
    "            skip_this_one=1\n",
    "            \n",
    "\n",
    "    print(np.nanmedian(np.abs(result2['Control Spear-Perterbation Spear']-np.nanmedian(result2['Control Spear-Perterbation Spear']))))\n",
    "    sns.violinplot(data=result2, x=Calculation, y='Condition',inner = \"quartile\")\n",
    "    print(ss2)\n",
    "    print(diff)\n",
    "    \n",
    "plt.xlim(-.05,.05)\n",
    "plt.xticks(np.arange(-.05,.05,.02))\n",
    "plt.savefig('Same_Expression_Sign_R2_Spear_Diff.pdf',bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "046f4d4d-f251-4299-836d-f55a8a05adf5",
   "metadata": {},
   "source": [
    "<h2> The analysis code used to quantify the correlations </h2> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3846485-9327-4442-9a62-0dcdb1d5838d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This is the code to quantify the correlations between genes on the computer cluster Biowulf\n",
    "#overall, it is not to complicated\n",
    "from npeet import entropy_estimators as ee\n",
    "import pandas as pd\n",
    "import os\n",
    "from multiprocessing import Pool\n",
    "import numpy as np\n",
    "from scipy.stats import norm, gaussian_kde\n",
    "from scipy.spatial.distance import pdist\n",
    "from scipy.spatial.distance import cdist\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import clear_output\n",
    "import time\n",
    "import copy\n",
    "import seaborn as sns\n",
    "from scipy.stats import entropy\n",
    "from pathlib import Path\n",
    "from math import nan\n",
    "import scipy\n",
    "from pathlib import Path\n",
    "from math import nan\n",
    "import signal\n",
    "from multiprocessing import Pool\n",
    "from scipy.stats import moment \n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.stats\n",
    "from scipy.misc import derivative\n",
    "from scipy.stats import moment \n",
    "import random\n",
    "\n",
    "\n",
    "\n",
    "#I want to quantify some MI metrics as well, thats what the following is for\n",
    "def MI_Single_Distances_2(v_1, v_2): \n",
    "\n",
    "    binst=[0,1,1000000000000]\n",
    "\n",
    "    temp1, temp2=np.histogram(v_1, bins=binst)\n",
    "\n",
    "    temp2, temp=np.histogram(v_2, bins=binst)\n",
    "\n",
    "    H, xedges, yedges = np.histogram2d(v_1, v_2, bins=(binst, binst))\n",
    "\n",
    "    MI=entropy(temp1, base=2)+entropy(temp2, base=2)-entropy(H.flatten(), base=2)\n",
    "\n",
    "    return MI\n",
    "\n",
    "def MI_Single_Distances_4(v_1, v_2): \n",
    "\n",
    "    binst_v1=[0,np.quantile(v_1, 0.25),np.quantile(v_1, 0.5), np.quantile(v_1, 0.75),1000000000]\n",
    "    binst_v2=[0,np.quantile(v_2, 0.25),np.quantile(v_2, 0.5), np.quantile(v_2, 0.75),1000000000]\n",
    "\n",
    "    temp1, temp2=np.histogram(v_1, bins=binst_v1)\n",
    "\n",
    "    temp2, temp=np.histogram(v_2, bins=binst_v2)\n",
    "\n",
    "    H, xedges, yedges = np.histogram2d(v_1, v_2, bins=(binst_v1, binst_v2))\n",
    "\n",
    "    MI=entropy(temp1, base=2)+entropy(temp2, base=2)-entropy(H.flatten(), base=2)\n",
    "\n",
    "    return MI\n",
    "\n",
    "\n",
    "def MI_Single_Distances_10(v_1, v_2): \n",
    "\n",
    "    binst_v1=[]\n",
    "    binst_v2=[]\n",
    "    for i in range(11):\n",
    "        binst_v1.append(np.quantile(v_1, 0.1*i))\n",
    "        binst_v2.append(np.quantile(v_2, 0.1*i))\n",
    "\n",
    "    temp1, temp2=np.histogram(v_1, bins=binst_v1)\n",
    "\n",
    "    temp2, temp=np.histogram(v_2, bins=binst_v2)\n",
    "\n",
    "    H, xedges, yedges = np.histogram2d(v_1, v_2, bins=(binst_v1, binst_v2))\n",
    "\n",
    "    MI=entropy(temp1, base=2)+entropy(temp2, base=2)-entropy(H.flatten(), base=2)\n",
    "\n",
    "    return MI\n",
    "\n",
    "\n",
    "\n",
    "def init_worker():\n",
    "    \"\"\"\n",
    "    This is necessary to be able to interrupt the\n",
    "    script with CTRL-C (or scancel for that matter).\n",
    "    It makes sure that the workers ignore SIGINT so\n",
    "    that any SIGINT sent goes to the master process\n",
    "    \"\"\"\n",
    "    signal.signal(signal.SIGINT, signal.SIG_IGN)\n",
    "\n",
    "\n",
    "def worker(kcat):\n",
    "    \n",
    "\n",
    "    try:\n",
    "\n",
    "        #Because this is going to be slow we are going to do one condition at a time.\n",
    "        Data_Folder='/data/bohrerch/RAD21d_T120/'\n",
    "        Data_Folder='/data/bohrerch/Temp/03072023/'\n",
    "        #Data_Folder='Temp/03072023/'\n",
    "\n",
    "        pick=np.random.randint(0,22)\n",
    "\n",
    "        \n",
    "        if pick==0:\n",
    "            condition='RAD21_ALL_1_'\n",
    "            condition_filename='RAD21d_T120/'\n",
    "            Specific_file_1='RAD21d_T120/HCT_DMSO2h.txt'\n",
    "            Specific_file_2='RAD21d_T120/HCT_AUXIN2h.txt'\n",
    "\n",
    "        if pick==1:\n",
    "            condition='RAD21_1_G1_'\n",
    "            condition_filename='RAD21d_T120/'\n",
    "            Specific_file_1='RAD21d_T120/HCT_DMSO2h_G1.txt'\n",
    "            Specific_file_2='RAD21d_T120/HCT_AUXIN2h_G1.txt'\n",
    "\n",
    "        if pick==2:\n",
    "            condition='JQ1_ALL_1_'\n",
    "            condition_filename='JQ1_T120/'\n",
    "            Specific_file_1='JQ1_T120/HCT_DMSOSr1.txt'\n",
    "            Specific_file_2='JQ1_T120/HCT_JQ1Sr1.txt'\n",
    "\n",
    "        if pick==3:\n",
    "            condition='JQ1_G1_1_'\n",
    "            condition_filename='JQ1_T120/'\n",
    "            Specific_file_1='JQ1_T120/HCT_DMSOSr1_G1.txt'\n",
    "            Specific_file_2='JQ1_T120/HCT_JQ1Sr1_G1.txt'\n",
    "\n",
    "        if pick==4:\n",
    "            condition='JQ1_All_2_'\n",
    "            condition_filename='JQ1_T120/'\n",
    "            Specific_file_1='JQ1_T120/HCT_DMSOSr2.txt'\n",
    "            Specific_file_2='JQ1_T120/HCT_JQ1Sr2.txt'\n",
    "\n",
    "        if pick==5:\n",
    "            condition='JQ1_G1_2_'\n",
    "            condition_filename='JQ1_T120/'\n",
    "            Specific_file_1='JQ1_T120/HCT_DMSOSr2_G1.txt'\n",
    "            Specific_file_2='JQ1_T120/HCT_JQ1Sr2_G1.txt'\n",
    "\n",
    "        if pick==6:\n",
    "            condition='MED_All_1_'\n",
    "            condition_filename='MED26d_T120/'\n",
    "            Specific_file_1='MED26d_T120/HCTMED26D_DMSOr1M.txt'\n",
    "            Specific_file_2='MED26d_T120/HCTMED26D_IAAr1M.txt'\n",
    "\n",
    "        if pick==7:\n",
    "            condition='MED_G1_1_'\n",
    "            condition_filename='MED26d_T120/'\n",
    "            Specific_file_1='MED26d_T120/HCTMED26D_DMSOr1M_G1.txt'\n",
    "            Specific_file_2='MED26d_T120/HCTMED26D_IAAr1M_G1.txt'\n",
    "\n",
    "        if pick==8:\n",
    "            condition='MED_All_2_'\n",
    "            condition_filename='MED26d_T120/'\n",
    "            Specific_file_1='MED26d_T120/HCTMED26D_DMSOr2M.txt'\n",
    "            Specific_file_2='MED26d_T120/HCTMED26D_IAAr2M.txt'\n",
    "\n",
    "        if pick==9:\n",
    "            condition='MED_G1_2_'\n",
    "            condition_filename='MED26d_T120/'\n",
    "            Specific_file_1='MED26d_T120/HCTMED26D_DMSOr2M_G1.txt'\n",
    "            Specific_file_2='MED26d_T120/HCTMED26D_IAAr2M_G1.txt'\n",
    "\n",
    "        if pick==10:\n",
    "            condition='Myc_All_1_'\n",
    "            condition_filename='Myc/'\n",
    "            Specific_file_1='Myc/xWTr1.txt'\n",
    "            Specific_file_2='Myc/Mycmr1.txt'\n",
    "\n",
    "        if pick==11:\n",
    "            condition='Myc_G1_1_'\n",
    "            condition_filename='Myc/'\n",
    "            Specific_file_1='Myc/xWTr1_G1.txt'\n",
    "            Specific_file_2='Myc/Mycmr1_G1.txt'\n",
    "\n",
    "        if pick==12:\n",
    "            condition='Myc_All_2_'\n",
    "            condition_filename='Myc/'\n",
    "            Specific_file_1='Myc/xWTr2.txt'\n",
    "            Specific_file_2='Myc/Mycmr2.txt'\n",
    "\n",
    "        if pick==13:\n",
    "            condition='Myc_G1_2_'\n",
    "            condition_filename='Myc/'\n",
    "            Specific_file_1='Myc/xWTr2_G1.txt'\n",
    "            Specific_file_2='Myc/Mycmr2_G1.txt'\n",
    "\n",
    "        if pick==14:\n",
    "            condition='rb_aB_All_1_'\n",
    "            condition_filename='rbnew/'\n",
    "            Specific_file_1='rbnew/CtrlSprB.txt'\n",
    "            Specific_file_2='rbnew/CtrlSpaB.txt'\n",
    "\n",
    "        if pick==15:\n",
    "            condition='rb_aB_G1_1_'\n",
    "            condition_filename='rbnew/'\n",
    "            Specific_file_1='rbnew/CtrlSprB_G1.txt'\n",
    "            Specific_file_2='rbnew/CtrlSpaB_G1.txt'\n",
    "\n",
    "        if pick==16:\n",
    "            condition='JQ1_R1_Control_to_JQ1_R2_Control_'\n",
    "            condition_filename='JQ1_T120_Controls/'\n",
    "            Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1.txt'\n",
    "            Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2.txt'\n",
    "    \n",
    "        if pick==17:\n",
    "            condition='JQ1_R1_G1_Control_to_JQ1_R2_G1_Control_'\n",
    "            condition_filename='JQ1_T120_Controls/'\n",
    "            Specific_file_1='JQ1_T120_Controls/HCT_DMSOSr1_G1.txt'\n",
    "            Specific_file_2='JQ1_T120_Controls/HCT_DMSOSr2_G1.txt'\n",
    "           \n",
    "        if pick==18:\n",
    "            condition='MED_R1_Control_to_MED_R2_Control_'\n",
    "            condition_filename='MED26d_T120_Controls/'\n",
    "            Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "            Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M.txt'\n",
    "\n",
    "        if pick==19:\n",
    "            condition='MED_R1_G1_Control_to_MED_R2_G1_Control_'\n",
    "            condition_filename='MED26d_T120_Controls/'\n",
    "            Specific_file_1='MED26d_T120_Controls/HCTMED26D_DMSOr1M_G1.txt'\n",
    "            Specific_file_2='MED26d_T120_Controls/HCTMED26D_DMSOr2M_G1.txt'\n",
    "            \n",
    "        #For accross conditions\n",
    "        if pick==20:\n",
    "            condition='MED_R1_Control_to_JQ1_R1_Control_'\n",
    "            condition_filename='MED_and_JQ1_Controls/'\n",
    "            Specific_file_1='MED_and_JQ1_Controls/HCTMED26D_DMSOr1M.txt'\n",
    "            Specific_file_2='MED_and_JQ1_Controls/HCT_DMSOSr1.txt'\n",
    "            \n",
    "        #For accross conditions\n",
    "        if pick==21:\n",
    "            condition='MED_R1_G1_Control_to_JQ1_R1_G1_Control_'\n",
    "            condition_filename='MED_and_JQ1_Controls/'\n",
    "            Specific_file_1='MED_and_JQ1_Controls/HCTMED26D_DMSOr1M_G1.txt'\n",
    "            Specific_file_2='MED_and_JQ1_Controls/HCT_DMSOSr1_G1.txt'\n",
    "        \n",
    "\n",
    "        #Lets load in the needed data: \n",
    "        print(Data_Folder+Specific_file_1)\n",
    "        print(Data_Folder+Specific_file_2)\n",
    "        \n",
    "        \n",
    "        #go through and organize the data\n",
    "        count=0\n",
    "        df_temp= pd.read_table(Data_Folder+Specific_file_1, skiprows=1, header=None, chunksize=1000)\n",
    "        for chunk in df_temp:\n",
    "            if count==0:\n",
    "                print('Go')\n",
    "                count=1\n",
    "                df=chunk\n",
    "            else:\n",
    "                df=pd.concat([df, chunk], ignore_index=True)\n",
    "       \n",
    "        count=0\n",
    "        df_temp= pd.read_table(Data_Folder+Specific_file_2, skiprows=1, header=None, chunksize=1000)\n",
    "        for chunk in df_temp:\n",
    "            if count==0:\n",
    "                print('Go2')\n",
    "                count=1\n",
    "                df2=chunk\n",
    "            else:\n",
    "                df2=pd.concat([df2, chunk], ignore_index=True)\n",
    "     \n",
    "        #Here we are going to normalize them the correct way\n",
    "        df=df.set_index(0)\n",
    "        df2=df2.set_index(0)\n",
    "        \n",
    "        Lowest_Expression_State=29347\n",
    "        df=df/np.nansum(df.values, axis=0)*Lowest_Expression_State\n",
    "        df2=df2/np.nansum(df2.values, axis=0)*Lowest_Expression_State\n",
    "        \n",
    "        \n",
    "        df=df.round(0)\n",
    "        df2=df2.round(0)\n",
    "        \n",
    "        #We are going to only look at genes that are expressed in at least 10% of the cells\n",
    "        genes_to_keep_control=df.index[np.nansum(df.values>0, axis=1)>len(df.iloc[0])/10]\n",
    "        genes_to_keep_perterb=df2.index[np.nansum(df2.values>0, axis=1)>len(df2.iloc[0])/10]\n",
    "        \n",
    "        #See how many genes we get to deal with\n",
    "        print(len(genes_to_keep_perterb))\n",
    "\n",
    "        #Here I am going to get rid of the genes that are not in both samples\n",
    "        values_to_kill=np.array([])\n",
    "        for kcat in df.index:\n",
    "            if kcat not in genes_to_keep_control.values:\n",
    "                values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "        df =df[~df.index.isin(values_to_kill)] \n",
    "        df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "\n",
    "\n",
    "        values_to_kill=np.array([])\n",
    "        for kcat in df2.index:\n",
    "            if kcat not in genes_to_keep_perterb.values:\n",
    "                values_to_kill=np.append(values_to_kill, kcat)\n",
    "\n",
    "        df =df[~df.index.isin(values_to_kill)]\n",
    "        df2 =df2[~df2.index.isin(values_to_kill)]  \n",
    "\n",
    "        try:\n",
    "\n",
    "            Gene_Locs_copy=pd.read_table(Data_Folder+condition_filename+\"Gene_Names\")\n",
    "            res,ind = np.unique(Gene_Locs_copy['name2'].values, return_index=True)\n",
    "            names_unique=res[np.argsort(ind)]\n",
    "\n",
    "        except:\n",
    "\n",
    "            names_unique=['Not there 1','Not there 2']\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        #We are going to generate a dataframe storing the geneomic distance as well as the correlation values. \n",
    "        for i in df.index.values[np.random.randint(0,len(df.index.values),len(df.index.values))]:\n",
    "\n",
    "            try:\n",
    "                file=Data_Folder+condition_filename+'Analyzed_Data/'+condition+i+'.csv'\n",
    "                df_temp= pd.read_table(file, skiprows=1, header=None, chunksize=100)\n",
    "                count=0\n",
    "                for chunk in df_temp:\n",
    "                    if count==0:\n",
    "                        count=1\n",
    "            except:\n",
    "                    #Reset the dataframe\n",
    "                    \n",
    "                    df_chromosome={'Binary MI Control':[], 'Binary MI Pert':[], '4 bin MI Control':[], '4 bin MI Pert':[],  '10 bin MI Control':[], '10 bin MI Pert':[],\n",
    "                                   'Spear Control':[], 'Spear Pert':[], 'True MI Control':[], 'True MI Pert':[], 'Exp Gene1 Control':[], 'Exp Gene2 Control':[] , 'Exp Gene1 Pert':[], 'Exp Gene2 Pert':[], \n",
    "                                   'Genomic Position gene 1':[], 'Genomic Position gene 2':[], 'Chrom Gene 1':[],'Chrom Gene 2':[]}\n",
    "                    \n",
    "                    df_chromosome=pd.DataFrame(df_chromosome)\n",
    "\n",
    "                    for ii in df.index.values:\n",
    "                        \n",
    "                        if i==ii:\n",
    "                            \n",
    "                            bad=1\n",
    "                            \n",
    "                        else:\n",
    "\n",
    "                                #All the correlation measures that we may use\n",
    "                                MI2_1=MI_Single_Distances_2(df.loc[df.index.values==i].values[0], df.loc[df.index.values==ii].values[0])\n",
    "                                MI2_2=MI_Single_Distances_2(df2.loc[df2.index.values==i].values[0], df2.loc[df2.index.values==ii].values[0])\n",
    "\n",
    "                                MI4_1=MI_Single_Distances_4(df.loc[df.index.values==i].values[0], df.loc[df.index.values==ii].values[0])\n",
    "                                MI4_2=MI_Single_Distances_4(df2.loc[df2.index.values==i].values[0], df2.loc[df2.index.values==ii].values[0])\n",
    "\n",
    "                                MI10_1=MI_Single_Distances_10(df.loc[df.index.values==i].values[0], df.loc[df.index.values==ii].values[0])\n",
    "                                MI10_2=MI_Single_Distances_10(df2.loc[df2.index.values==i].values[0], df2.loc[df2.index.values==ii].values[0])\n",
    "\n",
    "                                S_1=scipy.stats.spearmanr(df.loc[df.index.values==i].values[0], df.loc[df.index.values==ii].values[0])\n",
    "                                S_2=scipy.stats.spearmanr(df2.loc[df2.index.values==i].values[0], df2.loc[df2.index.values==ii].values[0])\n",
    "\n",
    "                                S_1=S_1[0]\n",
    "                                S_2=S_2[0]\n",
    "\n",
    "                                MI_1=ee.mi(df.loc[df.index.values==i].values[0], df.loc[df.index.values==ii].values[0])\n",
    "                                MI_2=ee.mi(df2.loc[df2.index.values==i].values[0], df2.loc[df2.index.values==ii].values[0])\n",
    "\n",
    "                                One=np.nanmean(df.loc[df.index.values==i].values[0])\n",
    "                                Two=np.nanmean(df.loc[df.index.values==ii].values[0])\n",
    "\n",
    "                                Three=np.nanmean(df2.loc[df2.index.values==i].values[0])\n",
    "                                Four=np.nanmean(df2.loc[df2.index.values==ii].values[0])\n",
    "\n",
    "                                try: \n",
    "\n",
    "                                    gen_1=Gene_Locs_copy['txStart'][Gene_Locs_copy['name2']==i].values\n",
    "                                    gen_2=Gene_Locs_copy['txStart'][Gene_Locs_copy['name2']==ii].values\n",
    "                                    gen_1=gen_1[0]\n",
    "                                    gen_2=gen_2[0]\n",
    "\n",
    "                                    chrom_1=Gene_Locs_copy['chrom'][Gene_Locs_copy['name2']==i].values\n",
    "                                    chrom_2=Gene_Locs_copy['chrom'][Gene_Locs_copy['name2']==ii].values\n",
    "                                    chrom_1=chrom_1[0]\n",
    "                                    chrom_2=chrom_2[0]\n",
    "\n",
    "                                except:\n",
    "                                    gen_1=nan\n",
    "                                    gen_2=nan\n",
    "                                    chrom_1=nan\n",
    "                                    chrom_2=nan\n",
    "\n",
    "                                df_chromosome.loc[len(df_chromosome.index)]=[MI2_1, MI2_2, MI4_1, MI4_2, MI10_1, MI10_2, S_1, S_2, MI_1, MI_2, One, Two, Three, Four, gen_1, gen_2, chrom_1, chrom_2]\n",
    "\n",
    "                    file=Data_Folder+condition_filename+'Analyzed_Data/'+condition+i+'.csv'\n",
    "                    df_chromosome.to_csv(file, index=False)\n",
    "                    print('Finished '+condition+i)\n",
    "\n",
    "    except: \n",
    "        \n",
    "        print('did not work')\n",
    "\n",
    "\n",
    "                                    \n",
    "                                    \n",
    "                                    \n",
    "if __name__ == '__main__':\n",
    "    \n",
    "    nproc = int(os.environ.get(\"SLURM_CPUS_PER_TASK\", \"16\"))\n",
    "    \n",
    "    print(\"Running on %d CPUs\" % nproc)\n",
    "    \n",
    "    p = Pool(nproc, init_worker)\n",
    "    \n",
    "    try:\n",
    "                # the the number of allocated cpus (or 2 if not run as a slurm job)\n",
    "                num_of_loci=1600\n",
    "                np.random.rand()\n",
    "                tasks = np.random.randint(0, 7591,1700)\n",
    "                \n",
    "                print('Hey Im starting') #Create a multiprocessing Pool\n",
    "                p.map(worker, tasks) \n",
    "\n",
    "    except (KeyboardInterrupt, SystemExit):\n",
    "            p.terminate()\n",
    "            p.join()\n",
    "            sys.exit(1)\n",
    "    else:\n",
    "            p.close()\n",
    "            p.join()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
